xbmc/cores/AudioEngine/Utils/PackerMAT.cpp

```diff
@@ -0,0 +1,440 @@
+
+      assert(remaining == 0);
+    }
+  }
+
+  // store the size of the current MAT frame, so we can add padding later
+  m_state.prevMatFramesize = m_state.matFramesize;
+  m_state.matFramesize = 0;
+
+  return true;
+}
+
+std::vector<uint8_t> CPackerMAT::GetOutputFrame()
+{
+  std::vector<uint8_t> buffer;
+
+  if (m_outputQueue.empty())
+    return {};
+
+  buffer = std::move(m_outputQueue.front());
+
+  m_outputQueue.pop_front();
+
+  return buffer;
+}
+
+void CPackerMAT::WriteHeader()
+{
+  m_buffer.resize(MAT_BUFFER_SIZE);
+
+  // reserve size for the IEC header and the MAT start code
+  const size_t size = BURST_HEADER_SIZE + mat_start_code.size();
+
+  // write MAT start code. IEC header written later, skip space only
+  memcpy(m_buffer.data() + BURST_HEADER_SIZE, mat_start_code.data(), mat_start_code.size());
+  m_bufferCount = size;
+
+  // unless the start code falls into the padding, it's considered part of the current MAT frame
+  // Note that audio frames are not always aligned with MAT frames, so we might already have a partial
+  // frame at this point
+  m_state.matFramesize += size;
+
+  // The MAT metadata counts as padding, if we're scheduled to write any, which mean the start bytes
+  // should reduce any further padding.
+  if (m_state.padding > 0)
+  {
+    // if the header fits into the padding of the last frame, just reduce the amount of needed padding
+    if (m_state.padding > size)
+    {
+      m_state.padding -= size;
+      m_state.matFramesize = 0;
+    }
+    else
+    {
+      // otherwise, consume all padding and set the size of the next MAT frame to the remaining data
+      m_state.matFramesize = (size - m_state.padding);
+      m_state.padding = 0;
+    }
+  }
+}
+
+void CPackerMAT::WritePadding()
+{
+  if (m_state.padding == 0)
+    return;
+
+  if (!m_logPadding && m_state.padding > MAT_BUFFER_SIZE / 2)
+  {
+    m_logPadding = true;
+    CLog::LogF(LOGWARNING,
+               "a large padding block of {} bytes is necessary due to unusual timestamps",
+               m_state.padding);
+  }
+  else if (m_logPadding && m_state.padding < MAT_BUFFER_SIZE / 2)
+    m_logPadding = false;
+
+  // for padding not writes any data (nullptr) as buffer is already zeroed
+  // only counts/skip bytes
+  const int remaining = FillDataBuffer(nullptr, m_state.padding, Type::PADDING);
+
+  // not all padding could be written to the buffer, write it later
+  if (remaining >= 0)
+  {
+    m_state.padding = remaining;
+    m_state.matFramesize = 0;
+  }
+  else
+  {
+    // more padding then requested was written, eg. there was a MAT middle/end marker
+    // that needed to be written
+    m_state.padding = 0;
+    m_state.matFramesize = -remaining;
+  }
+}
+
+void CPackerMAT::AppendData(const uint8_t* data, int size, Type type)
+{
+  // for padding not write anything, only skip bytes
+  if (type == Type::DATA)
+    memcpy(m_buffer.data() + m_bufferCount, data, size);
```

Comment:
Can you update just use m_buffer.size() instead of tracking it in m_bufferCount?

Reply:
Is for performance reasons: this works at very high rate 1200 times/second and I want avoid resize vector every ~0.83 ms

Vector is used only as buffer for allocate memory and easy transfer latter to `std::deque`. 

Vector "only" is resized one time every ~20 ms instead to clean and fill zeros.

Second reason is that is not filled sequentially (padding bytes are skipped): technically fist is filled with zeros and after are written only some non-contiguous areas.

Reply:
Ah ok, would it possible to reserve an amount to avoid reallocations?

Reply:
Already done it in L47:

```
CPackerMAT::CPackerMAT()
{
  m_buffer.reserve(MAT_BUFFER_SIZE);
}
```